### PR TITLE
Shared Loader - Add support for more rewriting points

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR's adds support for adding more than only once rewriting points in the loader.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Benchmarks Report :snail:

Benchmarks for #2487 compared to [master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d): 
* 1 benchmarks are **faster**, with geometric mean 1.146
* 1 benchmarks are **slower**, with geometric mean 1.133
* All benchmarks have **the same** allocations


The following thresholds were used for comparing the benchmark speeds:
* Mann–Whitney U test with statistical test for significance of **5%**
* Only results indicating a difference greater than **10%** and **0.3 ns** are considered.

Allocation changes below **0.5%** are ignored.


### Benchmark details


<details><summary><strong>Benchmarks.Trace.AgentWriterBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`WriteAndFlushEnrichedTraces`|net472|894μs|2.9μs|10.9μs|0| 0|0|3.16&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|711μs|3.41μs|14.1μs|0| 0|0|2.57&nbsp;KB|
|#2487|`WriteAndFlushEnrichedTraces`|net472|913μs|3.77μs|14.1μs|0| 0|0|3.16&nbsp;KB|
|#2487|`WriteAndFlushEnrichedTraces`|netcoreapp3.1|705μs|2.67μs|9.99μs|0| 0|0|2.57&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.AspNetCoreBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendRequest`|net472|0ns|0ns|0ns|0| 0|0|0&nbsp;b|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendRequest`|netcoreapp3.1|339μs|1.62μs|6.69μs|0.168| 0|0|19.71&nbsp;KB|
|#2487|`SendRequest`|net472|0ns|0ns|0ns|0| 0|0|0&nbsp;b|
|#2487|`SendRequest`|netcoreapp3.1|350μs|1.12μs|4.33μs|0.172| 0|0|19.71&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.DbCommandBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`ExecuteNonQuery`|net472|1.75μs|8.83ns|37.4ns|0.0923| 0.000855|0|586&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`ExecuteNonQuery`|netcoreapp3.1|1.62μs|9.24ns|69.1ns|0.00857| 0|0|624&nbsp;B|
|#2487|`ExecuteNonQuery`|net472|1.86μs|10.6ns|79.1ns|0.0929| 0|0|586&nbsp;B|
|#2487|`ExecuteNonQuery`|netcoreapp3.1|1.63μs|4.09ns|15.8ns|0.00895| 0|0|624&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.ElasticsearchBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`CallElasticsearch`|net472|2.54μs|12.8ns|54.2ns|0.126| 0|0|794&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`CallElasticsearch`|netcoreapp3.1|1.89μs|10.3ns|57.5ns|0.0106| 0|0|784&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`CallElasticsearchAsync`|net472|2.95μs|14.2ns|55.1ns|0.147| 0|0|931&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`CallElasticsearchAsync`|netcoreapp3.1|1.94μs|11.3ns|93.7ns|0.0125| 0|0|904&nbsp;B|
|#2487|`CallElasticsearch`|net472|2.79μs|13.8ns|60.3ns|0.126| 0|0|794&nbsp;B|
|#2487|`CallElasticsearch`|netcoreapp3.1|1.9μs|11ns|88.6ns|0.0108| 0|0|784&nbsp;B|
|#2487|`CallElasticsearchAsync`|net472|3.02μs|13.9ns|53.8ns|0.148| 0.00154|0|931&nbsp;B|
|#2487|`CallElasticsearchAsync`|netcoreapp3.1|1.94μs|11.3ns|99.4ns|0.0122| 0|0|904&nbsp;B|
</details>


<details><summary><strong>Benchmarks.Trace.GraphQLBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`ExecuteAsync`|net472|3.05μs|9.81ns|36.7ns|0.167| 0.00152|0|1.05&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`ExecuteAsync`|netcoreapp3.1|2.14μs|7.16ns|27.7ns|0.014| 0|0|1.02&nbsp;KB|
|#2487|`ExecuteAsync`|net472|3.17μs|13.5ns|50.3ns|0.166| 0.00158|0|1.05&nbsp;KB|
|#2487|`ExecuteAsync`|netcoreapp3.1|2.27μs|8.59ns|33.3ns|0.0145| 0|0|1.02&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.HttpClientBenchmark</strong> - Slower :warning: Same allocations :heavy_check_mark:



#### Slower :warning: in #2487

|  Benchmark         | diff/base | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.HttpClientBenchmark.SendAsync&#8209;netcoreapp3.1 | 1.133 | 4,898.91 | 5,552.85 | 

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendAsync`|net472|6.86μs|18.5ns|69.2ns|0.363| 0|0|2.3&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendAsync`|netcoreapp3.1|4.91μs|19.9ns|74.5ns|0.0298| 0|0|2.2&nbsp;KB|
|#2487|`SendAsync`|net472|7.33μs|12.7ns|45.9ns|0.364| 0|0|2.3&nbsp;KB|
|#2487|`SendAsync`|netcoreapp3.1|5.58μs|29.8ns|152ns|0.0296| 0|0|2.2&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.ILoggerBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|net472|3.36μs|17.6ns|88.2ns|0.225| 0|0|1.44&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|netcoreapp3.1|3.22μs|14.8ns|57.4ns|0.0204| 0|0|1.52&nbsp;KB|
|#2487|`EnrichedLog`|net472|3.43μs|11.8ns|47.3ns|0.225| 0|0|1.44&nbsp;KB|
|#2487|`EnrichedLog`|netcoreapp3.1|3.3μs|7.08ns|24.5ns|0.02| 0|0|1.52&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.Log4netBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|net472|303μs|1.73μs|13.1μs|0.447| 0.149|0|4.31&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|netcoreapp3.1|233μs|697ns|2.61μs|0| 0|0|4.19&nbsp;KB|
|#2487|`EnrichedLog`|net472|304μs|1.14μs|4.4μs|0.456| 0.152|0|4.31&nbsp;KB|
|#2487|`EnrichedLog`|netcoreapp3.1|245μs|613ns|2.21μs|0| 0|0|4.19&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.NLogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|net472|7.67μs|32.2ns|129ns|0.504| 0|0|3.21&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|netcoreapp3.1|6.68μs|38.3ns|281ns|0.0481| 0|0|3.58&nbsp;KB|
|#2487|`EnrichedLog`|net472|8.24μs|47.9ns|415ns|0.503| 0|0|3.21&nbsp;KB|
|#2487|`EnrichedLog`|netcoreapp3.1|7.06μs|30.3ns|117ns|0.0488| 0|0|3.58&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.RedisBenchmark</strong> - Faster :tada: Same allocations :heavy_check_mark:



#### Faster :tada: in #2487

|  Benchmark         | base/diff | Base Median (ns) | Diff Median (ns) | Modality   |
|:----------|-----------:|-----------:|--------:|--------:|
| Benchmarks.Trace.RedisBenchmark.SendReceive&#8209;net472 | 1.146 | 2,850.25 | 2,487.26 | several?

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendReceive`|net472|2.87μs|16.6ns|146ns|0.156| 0|0|1&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`SendReceive`|netcoreapp3.1|2.08μs|8.72ns|33.8ns|0.0133| 0|0|1&nbsp;KB|
|#2487|`SendReceive`|net472|2.51μs|13.4ns|72.3ns|0.159| 0.00134|0|1&nbsp;KB|
|#2487|`SendReceive`|netcoreapp3.1|2.18μs|10.2ns|41ns|0.0141| 0|0|1&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SerilogBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|net472|6μs|30.1ns|131ns|0.288| 0|0|1.85&nbsp;KB|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`EnrichedLog`|netcoreapp3.1|5.34μs|20.3ns|75.8ns|0.0189| 0|0|1.47&nbsp;KB|
|#2487|`EnrichedLog`|net472|6.23μs|26.9ns|101ns|0.288| 0|0|1.85&nbsp;KB|
|#2487|`EnrichedLog`|netcoreapp3.1|5.43μs|18.5ns|69.4ns|0.0188| 0|0|1.47&nbsp;KB|
</details>


<details><summary><strong>Benchmarks.Trace.SpanBenchmark</strong> - Same speed :heavy_check_mark: Same allocations :heavy_check_mark:

</summary>

### Raw results


|  Branch |                      Method |     Toolchain |     Mean | StdError |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------|---------------------------- |-------------- |---------:|---------:|--------:|-------:|------:|------:|----------:|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`StartFinishSpan`|net472|926ns|4.8ns|23.5ns|0.0707| 0|0|449&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`StartFinishSpan`|netcoreapp3.1|898ns|3.76ns|14.5ns|0.00593| 0|0|448&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`StartFinishScope`|net472|1.11μs|5.91ns|45.8ns|0.0832| 0|0|530&nbsp;B|
|[master](https://github.com/DataDog/dd-trace-dotnet/tree/4fdd8a8bd107cd36bbaad62d803030f5e93a492d)|`StartFinishScope`|netcoreapp3.1|1.08μs|5.17ns|20.7ns|0.00793| 0|0|568&nbsp;B|
|#2487|`StartFinishSpan`|net472|952ns|2.01ns|7.25ns|0.0704| 0|0|449&nbsp;B|
|#2487|`StartFinishSpan`|netcoreapp3.1|930ns|3.52ns|13.2ns|0.00605| 0|0|448&nbsp;B|
|#2487|`StartFinishScope`|net472|1.14μs|5.14ns|18.5ns|0.0831| 0|0|530&nbsp;B|
|#2487|`StartFinishScope`|netcoreapp3.1|1.21μs|6.35ns|33ns|0.00816| 0|0|568&nbsp;B|
</details>

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #2487 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/d5e6560757253da38d8b2557a8eae4edee277abf) will **not change** line coverage
:heavy_check_mark: Merging #2487 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/d5e6560757253da38d8b2557a8eae4edee277abf) will **not change** branch coverage
:heavy_check_mark: Merging #2487 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/d5e6560757253da38d8b2557a8eae4edee277abf) will **not change** complexity


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/d5e6560757253da38d8b2557a8eae4edee277abf) | #2487       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `11633` / `15979` | `11631` / `15979` |          |
| Lines %   | `73%`      | `73%`      |  `0%` :heavy_check_mark:  |
| Branches  | `6248` / `9348` | `6249` / `9348` |          |
| Branches %| `67%`      | `67%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `10558`      | `10558`        |  `0`  :heavy_check_mark:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=91398&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/91398/artifacts?artifactName=Code%20Coverage%20Report_91398&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2487](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=91387&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/91387/artifacts?artifactName=Code%20Coverage%20Report_91387&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/d5e6560757253da38d8b2557a8eae4edee277abf) | #2487       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `73%`      | `73%`       |  `0%` :heavy_check_mark:  |
| Branches %| `67%`      | `67%`       |  `0%` :heavy_check_mark:  |
| Complexity| `10558`      | `10558`       |  `0` :heavy_check_mark:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.ClrProfiler.AutoInstrumentation.MongoDb.MongoDbIntegration](https://github.com/DataDog/dd-trace-dotnet/tree/929ee14f865854125dd2585521d0623ddefb2033/tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/MongoDb/MongoDbIntegration.cs) | `-7%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=91398&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/91398/artifacts?artifactName=Code%20Coverage%20Report_91398&api-version=6.0&%24format=zip)  
* [HTML report for this PR #2487](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=91387&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/91387/artifacts?artifactName=Code%20Coverage%20Report_91387&api-version=6.0&%24format=zip)

**Review by gleocadie — APPROVED** [review]
LGTM

**Review by kevingosse — APPROVED** [review]

File: shared/src/native-src/loader.cpp [hunk 1]
```diff
@@ -322,52 +328,97 @@ namespace shared
...
+                const auto& specificAssemblyInjectName = std::get<0>(specMethodPair);
+                if (specificAssemblyInjectName != assemblyNameString)
+                {
+                    continue;
+                }
+
+                const auto& specificTypeToInjectName = std::get<1>(specMethodPair);
+                const auto& specificMethodToInjectName = std::get<2>(specMethodPair);
+
                 mdTypeDef appDomainTypeDef;
```
Comment (zacharycmontoya) [inline shared/src/native-src/loader.cpp:353]: nit: I think this variable could be renamed now

Comment (tonyredondo) [reply shared/src/native-src/loader.cpp:353]: oh you're right, I will change it.

Comment (tonyredondo) [reply shared/src/native-src/loader.cpp:353]: https://github.com/DataDog/dd-trace-dotnet/pull/2504

**Pull request merged**